### PR TITLE
Photon/P1: TCPClient: non-blocking, blocking, blocking with timeout writes support

### PR DESCRIPTION
**Pull request by avtolstoy**
### Problem

Unlike `TCPClient::read()`, `TCPClient::write()` is a blocking call and whenever TCP send queue gets completely filled (see #1461 for a particular case on Photon/P1 with low wifi signal), the application thread may be blocked for an indefinite amount of time.

There needs to be a way to specify the maximum amount of time the write should take.

### Solution

New `TCPClient::write()` overloads are added:
- `size_t write(uint8_t b, system_tick_t timeout);`
- `size_t write(const uint8_t *buf, size_t size, system_tick_t timeout);`

### Additional notes

`write()` methods return `size_t`, which is an unsigned integer, however in case of an error, `TCPClient::write()` method may return a negative error code cast to `size_t`. We are still keeping this behavior, with a thought of deprecating it in the next release.

Applications should use `TCPClient::getWriteError()/clearWriteError()` methods to check for errors, while `TCPClient::write()` should only return the number of bytes written.

### Steps to Test

- `wiring/api`

### Example App

```c++
#include "application.h"

SYSTEM_MODE(SEMI_AUTOMATIC);

Serial1LogHandler dbg(115200, LOG_LEVEL_ALL);

TCPServer tcpServerNonBlocking(5000);
TCPServer tcpServerBlockingTimeout(5001);
TCPServer tcpServerBlocking(5555);

void setup() {
    WiFi.on();
    WiFi.connect();
    WiFi.selectAntenna(ANT_EXTERNAL);
}

static void sendBunchOfData(TCPClient&& client, system_tick_t timeout) {
    while (client.connected()) {
        uint8_t buf[500];
        memset(buf, 'a', 500);
        int rc = client.write(buf, sizeof(buf), timeout);
        LOG(TRACE, "%d", rc);
    }
}

void loop() {
    sendBunchOfData(tcpServerNonBlocking.available(), 0);
    sendBunchOfData(tcpServerBlockingTimeout.available(), 1000);
    sendBunchOfData(tcpServerBlocking.available(), SOCKET_WAIT_FOREVER);
}
```

Connect to the appropriate port using telnet or something similar, while looking at the log place a finger onto the external antenna connector. Device should be unable to send data, causing TCP send queue to fill up. The behavior on each of the ports should be the following:
- 5000: as soon as the tcp queue is full, the log should be filled with a constant flow of -16 error codes
- 5001: as soon as the tcp queue is full, the log should show -16 error codes every second
- 5555: as soon as the tcp queue is full, there should be no log entries, indicating that the application thread is blocked

### References

- #1461
- particle-iot/photon-wiced#25
- [CH10639]

---

### Completeness

- [x] User is totes amazing for contributing!
- [x] Contributor has signed CLA ([Info here](https://github.com/spark/firmware/blob/develop/CONTRIBUTING.md))
- [x] Problem and Solution clearly stated
- [ ] Run unit/integration/application tests on device
- [ ] Added documentation **NEEDS TO BE UPDATED**
- [ ] Added to CHANGELOG.md after merging (add links to docs and issues)

### CODE REVIEW

File: wiring/inc/spark_wiring_print.h [hunk 1]
```diff
@@ -66,6 +67,9 @@ class Print
     }
     virtual size_t write(const uint8_t *buffer, size_t size);
 
+    virtual size_t write(uint8_t b, system_tick_t timeout) { return write(b); }
```
Comment (m-mcgowan) [inline wiring/inc/spark_wiring_print.h]: I would prefer to be careful here and add the new `write` methods to TCPClient only. The reason is that there is only one stream implementation that supports it, and I imagine users will be confused with the presence of the new methods that don't do anything useful for most streams.

File: wiring/src/spark_wiring_tcpclient.cpp [hunk 2]
```diff
@@ -110,12 +110,30 @@ int TCPClient::connect(IPAddress ip, uint16_t port, network_interface_t nif)
...
+}
+
+size_t TCPClient::write(uint8_t b, system_tick_t timeout)
+{
+    return write(&b, 1, timeout);
+}
+
+size_t TCPClient::write(const uint8_t *buffer, size_t size, system_tick_t timeout)
+{
+    int ret = status() ? socket_send_ex(d_->sock, buffer, size, 0, timeout, nullptr) : -1;
```
Comment (m-mcgowan) [inline wiring/src/spark_wiring_tcpclient.cpp]: should we call `clearWriteError()` here?

**Comment by elcojacobs** [comment]
How would you feel about being able to set a timeout value on the TcpClient class instance, instead of passing it with every function call?
That avoids the function overloading and I think will be even easier to use.

**Comment by avtolstoy** [comment]
Sounds like a good idea analogous to how it's done with regular BSD-socket interface: `setsockopt()` with `SO_RCVTIMEO` and `SO_SNDTIMEO`. Having both options available also doesn't sound too bad :)

I also agree that `size_t write(const char *str, system_tick_t timeout)` needs to be added. It was an oversight on my part.

**Comment by elcojacobs** [comment]
Are you going to implement this? I can have a go at it too if you want.

**Pull request merged**